### ISSUE TITLE
Configuration options for ApprovalTests

### ISSUE THREAD

**Issue by maxalbert** [approvals/ApprovalTests.Python#9]
It would be nice if it was possible to configure the behaviour of `ApprovalTests`.

One concrete use case I have is that I'd like to be able to store the `*.approved.txt` files in a subdirectory (say, `approved_files/`) rather than having them alongside the test files, which looks very cluttered and makes it hard to see what's going on. I presume there are other things for which configuration options would also be nice.

Does the original `ApprovalTests` (i.e., the non-Python version) have a way of specifying configuration options? I suspect it should be easy to point `ApprovalTests` to a custom configuration file (and/or make it read a configuration file with a standard name by default). I'm happy to have a shot at implementing this but thought I'd ask whether there is already anything that I missed because it's not documented, and also whether you have any ideas how it should ideally behave.

Many thanks!

**Comment by tdpreece** [comment]
I'm not sure about ApprovalTests in other languages but I think this is a feature that I would like too.  
How were you thinking of specifying the directory?  An environment variable?

**Comment by jamesrcounts** [comment]
Yes, this feature is available in at least 3 other ports of Approvals (.net, Perl, and JavaScript).  So its a common feature request and a pretty good one.  .Net uses a code annotation to specify the directory, JavaScript uses a configuration blob which can be specified in-line or stored as an external file in the user's home directory, and Perl uses a module-level variable.  My preference would be to start with a config file in a convention based location.  Similar to how GenericDiffReporterFactory finds it's config file.  

Home directory would be fine, but the problem with home directories (and environment variables) is that these locations aren't source controlled.  Ideally we want to have the option to checkin our approval config with the project so that we don't have to reconfigure every machine. (this was an actual problem with one team I worked with and they ended up writing code in test setup in order to set the environment variables).  Later down the line we can add additional conventions to tell approvals to look at an environment variable or in the home directory, but for now I'd say keep the config file somewhere near the tests.

**Comment by maxalbert** [comment]
@tdpreece An environment variable could be an option, but it is likely that users will want additional configuration options in the future, in which case multiple environment variables quickly become cumbersome. Also, it should be possible to set this option on a per-project basis, which is not possible with a global environment variable.

I was thinking that a simple config file could be convenient. For example, we could have a file called `approvaltests.cfg` (and/or `.approvaltests.cfg` if the user prefers to have it hidden by default) which might look something like this:
```
approved_files_dir = ./approved_files/
```
What I'm slightly unsure about is where this file should live (toplevel of the project, or the tests directory, or somewhere else?) and how `ApprovalTests` would detect it. I guess I'll have to play with it a bit more to be able to make a good suggestion. Do you have any ideas?

**Comment by maxalbert** [comment]
@jamesrcounts I just saw your reply as well. Thanks for the quick feedback from both of you. :) Yes, I think keeping it close to the tests is a good idea - in particular because the user may want to use different values for different projects, so for this particular option a global config file in the user's home directory doesn't sound quite right (although I'm sure there are sensible global options for which it would make sense - so it would likely be good to support both a global config file and a local one, with options in the local one overriding the global ones).

I can also imagine that similar to the .Net and Java/JavaScript versions it could be nice to use decorators or with statements in Python to allow configuring this inline, but maybe it's best to keep it simple in the beginning and go with a config file until there is an actual need/user request for this.

Out of interest, what additional configuration options do the other ports of ApprovalTests provide?

**Comment by jamesrcounts** [comment]
use JSON as the configuration language.  I chose JSON for the reporter file because in the long run I'd like the different ports to be able to standardize and share configurations with each other.  

I think in the past I've solved the "where" problem by looking in the test directory, then start climbing up the tree until i found the config.  This has the advantage of allowing custom config for certain test suites and a global config for the project, or even multiple projects if you allow the search to climb out side the project's own folders.

The most common configuration options are: Where to put the approval files.  What reporter to use.  Whether to consider differences in line endings (`\n` vs `\r\n`) as actual differences.  Javascript readme was pretty good documentation of its options.
